`hdr_value_at_percentile` optimization.

yoav-steinberg opened the pull request:
This is basically a C port of the optimization in PR https://github.com/HdrHistogram/hdrhistogram-go/pull/48 by @filipecosta90.

before:
```
yoav@yoav-thinkpad:~/HdrHistogram_c/build$ ./test/hdr_histogram_benchmark --benchmark_filter=BM_hdr_value_at_percentile/3/86400000 --benchmark_min_time=10
2022-03-21 12:41:46
Running ./test/hdr_histogram_benchmark
Run on (12 X 4600 MHz CPU s)
CPU Caches:
  L1 Data 32K (x6)
  L1 Instruction 32K (x6)
  L2 Unified 256K (x6)
  L3 Unified 12288K (x1)
Load Average: 0.68, 0.81, 0.89
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
***WARNING*** Library was built as DEBUG. Timings may be affected.
-----------------------------------------------------------------------------------
Benchmark                                         Time             CPU   Iterations
-----------------------------------------------------------------------------------
BM_hdr_value_at_percentile/3/86400000         34937 ns        34955 ns       404859
BM_hdr_value_at_percentile/3/86400000000      36561 ns        36581 ns       398987
```
after:
```
yoav@yoav-thinkpad:~/HdrHistogram_c/build$ ./test/hdr_histogram_benchmark --benchmark_filter=BM_hdr_value_at_percentile/3/86400000 --benchmark_min_time=10
2022-03-21 13:36:17
Running ./test/hdr_histogram_benchmark
Run on (12 X 4600 MHz CPU s)
CPU Caches:
  L1 Data 32K (x6)
  L1 Instruction 32K (x6)
  L2 Unified 256K (x6)
  L3 Unified 12288K (x1)
Load Average: 0.86, 0.96, 1.26
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
***WARNING*** Library was built as DEBUG. Timings may be affected.
-----------------------------------------------------------------------------------
Benchmark                                         Time             CPU   Iterations
-----------------------------------------------------------------------------------
BM_hdr_value_at_percentile/3/86400000          5070 ns         5083 ns      2795009
BM_hdr_value_at_percentile/3/86400000000       5124 ns         5135 ns      2753732
```

filipecosta90 commented:
great improvement @yoav-steinberg! thank you! 
@mikeb01 do you think this PR can be reviewed? In a nutshell, it avoids wasteful computation on the iterator and just computes the bare minimum up until we reach the value at percentile.

mikeb01 commented:
This definitely looks interesting.  Which build configuration was used to generate the benchmarks Release or Debug.  If this was Debug, it would be good to rerun them with a release build.

mikeb01 commented on src/hdr_histogram.c:
```diff
@@ -666,29 +675,40 @@ int64_t hdr_min(const struct hdr_histogram* h)
+    int32_t bucket_base_idx = get_bucket_base_index(h, bucket_idx);
+
+    for (;;) {
+        if (count_to_idx >= count_at_percentile) {
```
I'm concerned about this being the only exit clause for the loop.  I think the sub_bucket_idx should also be checked to make sure that it doesn't run of the end or get into an infinite loop.

yoav-steinberg replied:
Taken from here: https://github.com/HdrHistogram/hdrhistogram-go/blob/494271c4c016b36c8cee88480288f33b419cf7b0/hdr.go#L350-L354

@filipecosta90 Can you check if there's an issue here?

filipecosta90 replied:
@mikeb01 isn't the following check and change enough?:
https://github.com/HdrHistogram/HdrHistogram_c/pull/107#diff-3b9d3dbfd2aefc19e97618e6ca54cdab9ec1900faef90b1a450acad891980717R694
notice when we cross the sub-bucket limit we change bucket:
```c
        if (sub_bucket_idx >= h->sub_bucket_count)
        {
            sub_bucket_idx = h->sub_bucket_half_count;
            bucket_idx++;
            bucket_base_idx = get_bucket_base_index(h, bucket_idx);
        }
```

yoav-steinberg replied:
@mikeb01 I think that since `count_at_percentile` is always less or equal to `total_count` then we can safely assume that `count_to_index` will reach or pass `count_at_percentile` before the call to `get_count_at_index_given_bucket_base_idx` is called with an overflowing `bucket_base_idx, sub_bucket_idx`. So I think it's safe.
@filipecosta90 please confirm.

yoav-steinberg replied:
I added an overflow check that'll abort early if the passed `count_at_percentile` is to high. Please re-review.

mikeb01 replied:
I think it works as long as all of the logic is correct and the total_count is consistent with the underlying data and nothing gets accidentally messed up.  "I think" being the operative term here.  My concern is that it takes a substantial mental effort and assumption of behaviour to verify correctness.  I would feel more confident with a simple check of the index against counts_len (the iterator move_next does this as an additional safety step).  I know this means a bit of restructuring of the code, e.g. in-lining `get_count_at_index_given_bucket_base_idx` so the index can be checked, but I think it will be safer and less fragile in the long run.

yoav-steinberg commented:
> Which build configuration was used...

I used _Release_ for the project itself: 
`cmake -DHDR_HISTOGRAM_BUILD_BENCHMARK=ON -DCMAKE_BUILD_TYPE=Release ..`
but couldn't get the benchmark lib to be built with Release, hence the:
`***WARNING*** Library was built as DEBUG. Timings may be affected.`
But I think it's still valid.

yoav-steinberg commented:
@mikeb01 @filipecosta90 I refactored and added an explicit `counts` bounds check as requested for extra safety. Note that added a slight performance hit, so I optimized a bit more to get back to the previous performance. I'd appreciate you verify that I didn't break the logic.
Thanks.

mikeb01 commented:
I've merged this, but rewrote the loop to match the implementation in the Java version.  The performance measured on my local machine was within 1%.
